ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -81,25 +89,29 @@ public static SourceBuilder instance() {
     final boolean forceMaterialization = !planInfo.isRepartitionedInPlan(streamSource);
 
     final KTable<K, GenericRow> transformed = source.transformValues(
-        new AddKeyAndPseudoColumns<>(keyGenerator, streamSource.getPseudoColumnVersion()));
+        new AddPseudoColumnsToMaterialize<>(streamSource.getPseudoColumnVersion()));
+
+    final KTable<K, GenericRow> toPotentiallyMaterialize;
 
     if (forceMaterialization) {
       // add this identity mapValues call to prevent the source-changelog
       // optimization in kafka streams - we don't want this optimization to
       // be enabled because we cannot require symmetric serialization between
       // producer and KSQL (see https://issues.apache.org/jira/browse/KAFKA-10179
       // and https://github.com/confluentinc/ksql/issues/5673 for more details)
-      return transformed.mapValues(row -> row, materialized);
-
+      toPotentiallyMaterialize = transformed.mapValues(row -> row, materialized);
     } else {
       // if we know this table source is repartitioned later in the topology,
       // we do not need to force a materialization at this source step since the
       // re-partitioned topic will be used for any subsequent state stores, in lieu
       // of the original source topic, thus avoiding the issues above.
       // See https://github.com/confluentinc/ksql/issues/6650
-      return transformed.mapValues(row -> row);
+      toPotentiallyMaterialize = transformed.mapValues(row -> row);
     }
 
+    return toPotentiallyMaterialize.transformValues(new AddRemainingPseudoAndKeyCols<>(
```

Comment:
The way that this is setup right now, when `forceMaterialization` is false we'll still have two transformValues() steps rather than one even though a single transformValues() would suffice. It's more performant to have a single one though it's also nice that the Streams topology doesn't vary based on whether we happen to be materializing or not. I could see us going either way on this decision... let's leave it as is for now and potentially revisit before releasing this feature.

Reply:
Sounds good, I'd be happy to discuss this as you see fit.